### PR TITLE
[Week 2]: Added Submit handler and GetBoards Modal

### PR DESCRIPTION
**Pull request by myselfmohdsameer**
## Proposed Changes

- Added submit handler support in Extend Configuration.
- Implemented submit handler
- Added dynamic boards fetching
- Designed get boards modal

### CODE REVIEW

File: MiroApp.ts [hunk 1]
```diff
@@ -115,4 +117,13 @@ export class MiroApp extends App {
             .getByUsername(this.botUsername)) as IUser;
         return true;
     }
+
+    public async executeViewSubmitHandler(context: UIKitViewSubmitInteractionContext, read: IRead, http: IHttp, persistence: IPersistence, modify: IModify) {
+        const handler = new ExecuteViewSubmitHandler(this, read, http, modify, persistence);
+        return await handler.run(this, context, read, http, persistence, modify);
```
Comment (tapiarafael) [inline MiroApp.ts:87]: Why pass all the params on the constructor and in the method? If you're already receiving these `(this, read, http, modify, persistence)` on the constructor, there's no need to pass again in the method call. You can pass only the `context`.

File: MiroApp.ts [hunk 2]
```diff
@@ -115,4 +117,13 @@ export class MiroApp extends App {
...
+      }
+    
+    protected async extendConfiguration(configuration: IConfigurationExtend): Promise<void> {
+    await Promise.all([this.getOauth2ClientInstance().setup(configuration), configuration.slashCommands.provideSlashCommand(new MiroCommand(this))]);
```
Comment (tapiarafael) [inline MiroApp.ts]:
```suggestion
        await Promise.all([this.getOauth2ClientInstance().setup(configuration), configuration.slashCommands.provideSlashCommand(new MiroCommand(this))]);
```

File: src/api/boards/getBoards.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import { IUIKitViewSubmitIncomingInteraction } from "@rocket.chat/apps-engine/definition/uikit/UIKitIncomingInteractionTypes";
+import { IUser } from "@rocket.chat/apps-engine/definition/users";
+import { getAccessTokenForUser } from "../../storage/users";
+import { ModalsEnum } from "../../enums/Modals";
+import { HttpStatusCode } from "@rocket.chat/apps-engine/definition/accessors";
+import { Block } from "@rocket.chat/ui-kit";
+import { getMiroBoardsUrl } from "../../lib/const";
+import { Texts } from "../../enums/Texts";
+
+export async function getBoards({ context, data, room, read, persistence, modify, http }: { context: UIKitViewSubmitInteractionContext; data: IUIKitViewSubmitIncomingInteraction; room: IRoom; read: IRead; persistence: IPersistence; modify: IModify; http: IHttp }) {
```
Comment (tapiarafael) [inline src/api/boards/getBoards.ts]:
```suggestion
interface IGetBoardsParams {
    context: UIKitViewSubmitInteractionContext; 
    data: IUIKitViewSubmitIncomingInteraction; 
    room: IRoom; 
    read: IRead; 
    persistence: IPersistence; 
    modify: IModify; 
    http: IHttp;
}

export async function getBoards({ context, data, room, read, persistence, modify, http }: IGetBoardsParams) {
```
It's easier to read and maintaining.

File: src/api/boards/getBoards.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    const builder = await modify.getCreator().startMessage().setRoom(room);
+    const block: Block[] = [];
+    // @TODO: handle response
+    await modify.getNotifier().notifyUser(user, builder.getMessage());
+  } else {
+    const textSender = await modify.getCreator().startMessage().setText(Texts.getBoardsFailure + response.data.err);
+    if (room) {
+      textSender.setRoom(room);
+    }
+    await modify.getCreator().finish(textSender);
```
Comment (tapiarafael) [inline src/api/boards/getBoards.ts:39]: Wouldn't this going to break if there's no room?

File: src/handlers/submitHandler.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        slashcommandcontext?: SlashCommandContext, 
+        uikitcontext?: UIKitInteractionContext) {
+		const { user, view } = context.getInteractionData();
+        const uiData = await getUIData(read.getPersistenceReader(), user.id);
+        const room = uiData.room;
+        const roomId  =  uiData.room?.id
+        const data = context.getInteractionData();    
+        try {
+            
+		switch (view.id) {
```
Comment (tapiarafael) [inline src/handlers/submitHandler.ts]: Careful with the indentation. Maybe we need to configure the prettier in the project.

File: src/handlers/submitHandler.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+	) {}
+
+	public async run(
+        app:MiroApp,
+        context: UIKitViewSubmitInteractionContext,
+        read: IRead,
+        http: IHttp,
+        persistence: IPersistence,
+        modify: IModify,
+        slashcommandcontext?: SlashCommandContext,
```
Comment (tapiarafael) [inline src/handlers/submitHandler.ts]: Do we need these `slashCommandContext` and `uiKitContext` params here? They're not being used anywhere.

File: src/modals/getBoardsModal.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
+import { IHttp, IModify, IPersistence, IRead, IUIKitSurfaceViewParam } from '@rocket.chat/apps-engine/definition/accessors';
+import { ModalsEnum } from '../enums/Modals';
+import { SlashCommandContext } from '@rocket.chat/apps-engine/definition/slashcommands';
+import { UIKitInteractionContext, UIKitSurfaceType } from '@rocket.chat/apps-engine/definition/uikit';
+import { Block } from '@rocket.chat/ui-kit';
+import { getActionsBlock, getButton, getInputBox, getOptions, getSectionBlock, getStaticSelectElement } from '../helpers/blockBuilder';
+
+export async function getBoardsModal({ modify, read, persistence, http, slashcommandcontext, uikitcontext, data }: { modify: IModify; read: IRead; persistence: IPersistence; http: IHttp; slashcommandcontext?: SlashCommandContext; uikitcontext?: UIKitInteractionContext; data?: string }): Promise<IUIKitSurfaceViewParam> {
```
Comment (tapiarafael) [inline src/modals/getBoardsModal.ts]: Same from the `api/boards/getBoards.ts`. Extract the parameters to an interface to be easier and better to maintain.

**Review by tapiarafael — CHANGES_REQUESTED** [review]

File: src/handlers/submitHandler.ts (hunk 6 above)
Comment (myselfmohdsameer) [reply src/handlers/submitHandler.ts]: Added for future scope.

File: src/modals/getBoardsModal.ts (hunk 7 above)
Comment (myselfmohdsameer) [reply src/modals/getBoardsModal.ts]: Noted.

File: src/api/boards/getBoards.ts (hunk 3 above)
Comment (myselfmohdsameer) [reply src/api/boards/getBoards.ts]: Sure sir, I will now move interfaces to a separate file and use them.

File: src/api/boards/getBoards.ts (hunk 4 above)
Comment (myselfmohdsameer) [reply src/api/boards/getBoards.ts:39]: I am persisting this via UIData in persistence.ts 
So it is very unlikely to fail.

**Comment by myselfmohdsameer** [comment]
@tapiarafael Sir I have resolved most comments with commit [bd9b179](https://github.com/RocketChat/Apps.Miro/pull/3/commits/bd9b179ecaa9107d97cfa9403298b8d4225ab57e)
PTAL, thanks.

File: MiroApp.ts (hunk 1 above)
Comment (myselfmohdsameer) [reply MiroApp.ts:87]: Adding changes in next PR

**Pull request merged**